src/core/net/srp_client.cpp

```diff
@@ -155,6 +152,83 @@ bool Client::Service::Matches(const Service &aOther) const
     return (strcmp(GetName(), aOther.GetName()) == 0) && (strcmp(GetInstanceName(), aOther.GetInstanceName()) == 0);
 }
 
+//---------------------------------------------------------------------
+// Client::AutoStart
+
+#if OPENTHREAD_CONFIG_SRP_CLIENT_AUTO_START_API_ENABLE
+
+Client::AutoStart::AutoStart(void)
+{
+    Clear();
+    mState = kDefaultMode ? kSelectedNone : kDisabled;
+}
+
+bool Client::AutoStart::HasSelectedServer(void) const
+{
+    bool hasSelected = false;
+
+    switch (mState)
+    {
+    case kDisabled:
+    case kSelectedNone:
+        break;
+
+    case kSelectedUnicastPreferred:
+    case kSelectedUnicast:
+    case kSelectedAnycast:
+        hasSelected = true;
+        break;
+    }
+
+    return hasSelected;
+}
+
+void Client::AutoStart::SetState(State aState)
+{
+    if (mState != aState)
+    {
+        LogInfo("AutoStartState %s -> %s", StateToString(mState), StateToString(aState));
+        mState = aState;
+    }
+}
+
+void Client::AutoStart::SetCallback(AutoStartCallback aCallback, void *aContext)
+{
+    mCallback = aCallback;
+    mContext  = aContext;
+}
+
+void Client::AutoStart::InvokeCallback(const Ip6::SockAddr *aServerSockAddr) const
+{
+    if (mCallback != nullptr)
+    {
+        mCallback(aServerSockAddr, mContext);
+    }
+}
+
+#if OT_SHOULD_LOG_AT(OT_LOG_LEVEL_INFO)
```

Comment:
We had some recent discussion on whether or not to include this check https://github.com/openthread/openthread/pull/7562#discussion_r845437116.

Regardless of whether we include or not, we should be consistent.

Thoughts?

Reply:
After some more thinking on this, here is my suggestion: 

- While we can rely on toolchain to remove unused methods (and more importantly their unused static const definitions), I feel it may be safer and better to not relay on toolchain behavior and take care of it the code. 
 - Specially for cases like here where the `ToString()` method is `private` and intended for use for logging within the same `class` so we should know what log level it is going to be used.
- So I prefer if we keep and/or re-add the `OT_SHOULD_LOG_AT()` guard checks.

Thoughts?

(when I get the chance, I am going to do some test with some `arm-gcc` toolchain to see if it does correctly remove unused const definitions).

Reply:
Sounds good to me. We can take this up in a separate issue/pr.